java/src/main/java/ai/rapids/cudf/ColumnView.java

```diff
@@ -2386,6 +2386,48 @@ public final ColumnVector stringReplace(Scalar target, Scalar replace) {
         replace.getScalarHandle()));
   }
 
+  /**
+   * For each string, replaces any character sequence matching the given pattern using the
+   * replacement string scalar.
+   *
+   * @param pattern The regular expression pattern to search within each string.
+   * @param repl The string scalar to replace for each pattern match.
+   * @return A new column vector containing the string results.
+   */
+  public final ColumnVector replaceRegex(String pattern, Scalar repl) {
+    return replaceRegex(pattern, repl, -1);
+  }
+
+  /**
+   * For each string, replaces any character sequence matching the given pattern using the
+   * replacement string scalar.
+   *
+   * @param pattern The regular expression pattern to search within each string.
+   * @param repl The string scalar to replace for each pattern match.
+   * @param maxRepl The maximum number of times a replacement should occur within each string.
+   * @return A new column vector containing the string results.
+   */
+  public final ColumnVector replaceRegex(String pattern, Scalar repl, int maxRepl) {
+    if (!repl.getType().equals(DType.STRING)) {
+      throw new IllegalArgumentException("Replacement must be a string scalar");
+    }
```

Comment:
What is the difference between throwing an exception and `assert type.equals(DType.STRING) : ....`? Should we throw exception so we can recover from it?

Reply:
Asserts can be removed at run time. Asserts are typically off by default in the JVM. This lets us put in some checks on the time critical data path, and then remove them in production when we are more confident that we will not hit that case. We have also done it in some cases when the underlying cudf code also checks for these cases, but we are not 100% sure that the cudf code is doing a redundant check. Because of this there are a lot of assets in the code that probably can be deleted because cudf has us covered.  There are also a number of asserts that probably should be a regular if/throw pattern.

In this case if we know that cudf is doing the check for us then we should just delete the check all together.

Reply:
They are _almost_ equivalent.  Asserts will throw exceptions but only if assertions are enabled (and they are not by default).  In this case, since we're blindly casting the native scalar handle to a `string_scalar` rather than a `scalar` I thought it prudent to always check rather than just when assertions are enabled.  Debugging a bad native cast is not fun.  In many of the other assertion cases, the native code will catch the mistake (e.g.: passing an integer column to a function expecting a string column) and throw an exception from native code.  The assertion in that case is just a bit easier to debug since it's coming straight from Java code.

Reply:
Oh and an assert can be recovered from. It throws an `AssertionError` instance and is essentially the equivalent of.

```
if (assertsEnabled) {
    if (!repl.getType().equals(DType.STRING)) {
      throw new AssertionError("Replacement must be a string scalar");
    }
}
```